src/v/cluster/topic_table.cc

```diff
@@ -647,23 +647,40 @@ void topic_table::notify_waiters() {
     /// If by invocation of this method there are no waiters, notify
     /// function_ptrs stored in \ref notifications, without consuming all
     /// pending_deltas for when a subsequent waiter does arrive
+
     std::vector<delta> changes;
+
+    // \ref notify_waiters is called after every apply. Hence for the most
+    // part there should only be a few items towards the end of \ref
+    // _pending_deltas that need to be sent to callbacks in \ref
+    // notifications. \ref _last_consumed_by_notifier_offset allows us to
+    // skip ahead to the items added since the last \ref notify_waiters
+    // call.
+    auto starting_iter = _pending_deltas.begin()
+                         + _last_consumed_by_notifier_offset;
+
     std::copy_if(
-      _pending_deltas.begin(),
+      starting_iter,
       _pending_deltas.end(),
       std::back_inserter(changes),
       [this](const delta& d) { return d.offset > _last_consumed_by_notifier; });
-    for (auto& cb : _notifications) {
-        cb.second(changes);
-    }
+
     if (!changes.empty()) {
+        for (auto& cb : _notifications) {
+            cb.second(changes);
+        }
+
         _last_consumed_by_notifier = changes.back().offset;
     }
 
+    _last_consumed_by_notifier_offset = _pending_deltas.end()
+                                        - _pending_deltas.begin();
+
     /// Consume all pending deltas
     if (!_waiters.empty()) {
         changes.clear();
         changes.swap(_pending_deltas);
+        _last_consumed_by_notifier_offset = 0;
```

Comment:
whenever there are waiters (common case?) the pending_deltas always gets cleared here after the swap. so i think i'm then a bit confused about when the optimization you introduced takes affect.

Reply:
I could be wrong, but the main thing we are trying to optimize is replying a controller log at start. We've seen in a recent incident reactor stalls in the existing `std::copy` section. This lead us to believe that there is a time before any waiters are registered where `_pending_deltas` can grow rather large. This optimization ensures that we are only iterating over a small subset of `_pending_deltas` that we have yet to send out to the registered callbacks.

Reply:
ok i think i'm seeing it more clearly today. so the issue is a combination of

1. pending deltas grows over time during controller replay
2. notify waiters is called frequently which re-processes all of the pending deltas
3. when there are no waiters, then pending_waiters never gets cleared

so the update here

```
    _last_consumed_by_notifier_offset = _pending_deltas.end()
                                        - _pending_deltas.begin();
```

ensures that the next time through only a partial set is notified? if so, makes sense!

Reply:
Yes, and it is (2) that makes it quadratic: this is caused by the `copy_if`, since it does O(n) work for n `_pending_deltas`. The rest of the method doesn't have this characteristic, since `changes` will have only O(1) size, amortized (i.e., every delta gets copied only once), but the copy from the start makes it quadratic.

Reply:
thanks!